### PR TITLE
HHH-15283 - fix NPE for NamedNativeQuery + SqlResultSetMapping (columns)

### PR DESCRIPTION
**Pull request by dreab8**
https://hibernate.atlassian.net/browse/HHH-15283

### CODE REVIEW

File: hibernate-core/src/test/java/org/hibernate/orm/test/query/resultmapping/NamedNativeQueryWithCountColumnTest.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+						entityManager.persist( sample );
+					}
+				}
+		);
+	}
+
+	@Test
+	void testNamedNativeQuery(EntityManagerFactoryScope scope) {
+		scope.inTransaction( em -> {
+			Long count = em.createNamedQuery( "sample.count", Long.class ).getSingleResult();
```
Comment (NathanQingyangXu) [inline hibernate-core/src/test/java/org/hibernate/orm/test/query/resultmapping/NamedNativeQueryWithCountColumnTest.java:53]: currently two dialects testing failed due to the above statement (Class casting exception). Seems some db returns different types. Now I saw this ticket is such an edge case.

File: hibernate-core/src/test/java/org/hibernate/orm/test/query/resultmapping/NamedNativeQueryWithCountColumnTest.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+			assertThat( count, is( 3l ) );
```
Comment (NathanQingyangXu) [inline hibernate-core/src/test/java/org/hibernate/orm/test/query/resultmapping/NamedNativeQueryWithCountColumnTest.java:54]: this is minor. But I think given the count is long type, `assertThat( count, is( 31L ) )` seems less confusing to me.

File: hibernate-core/src/main/java/org/hibernate/query/results/complete/CompleteResultBuilderBasicValuedStandard.java [hunk 3]
```diff
@@ -166,14 +166,14 @@ public boolean equals(Object o) {
...
+		return Objects.equals( explicitJavaType, that.explicitJavaType );
 	}
 
 	@Override
 	public int hashCode() {
 		int result = explicitColumnName != null ? explicitColumnName.hashCode() : 0;
 		result = 31 * result + ( explicitType != null ? explicitType.hashCode() : 0 );
-		result = 31 * result + explicitJavaType.hashCode();
+		result = 31 * result + ( explicitJavaType != null ? explicitJavaType.hashCode() : 0 );
 		return result;
```
Comment (NathanQingyangXu) [inline hibernate-core/src/main/java/org/hibernate/query/results/complete/CompleteResultBuilderBasicValuedStandard.java:177]: My initial change is to use Objects util methods aggressively as following:
```
		return Objects.equals( explicitColumnName, that.explicitColumnName )
				&& Objects.equals( explicitType, that.explicitType )
				&& Objects.equals( explicitJavaType, that.explicitJavaType );
```
and
```
return Objects.hash( explicitColumnName, explicitType, explicitJavaType );
```

but later on I reverted back the above changes for I think maybe you have some special reason to avoid the approach (consistent and error-free to me) for performance reason?

**Comment by NathanQingyangXu** [comment]
I just realized the PR title might be imperfect. The condition is not normal real column (not tested though) but seudo column like 'count(*)', so renaming the title as "fix NPE for NamedNativeQuery + SqlResultSetMapping (seudo column like 'count(*)').

File: hibernate-core/src/test/java/org/hibernate/orm/test/query/resultmapping/NamedNativeQueryWithCountColumnTest.java (hunk 1 above)
Comment (dreab8) [reply hibernate-core/src/test/java/org/hibernate/orm/test/query/resultmapping/NamedNativeQueryWithCountColumnTest.java]: restricted to h2
